File: Growler/ProductListViewController.m
//
//  MasterViewController.m
//  Growler
//
//  Created by Shopify.
//  Copyright (c) 2015 Shopify Inc. All rights reserved.
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in
//  all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
//  THE SOFTWARE.
//

#import <Buy/BUYClient.h>
#import "Credentials.h"
#import "ProductListViewController.h"
#import "ProductViewController.h"
#import "Theme.h"
#import "ShippingRatesTableViewController.h"
#import "Growler-Swift.h"
#import "AsyncImageView.h"

#import <Buy/Buy.h>
#import <Buy/BUYClient+Storefront.h>

@interface ProductListViewController () /*<UIViewControllerPreviewingDelegate>*/

@property (nonatomic, strong) BUYClient *client;
@property (nonatomic, strong) BUYCollection *collection;
@property (nonatomic, strong) NSOperation *collectionOperation;

@end

@implementation ProductListViewController

@synthesize products = _products;

- (instancetype)initWithClient:(BUYClient *)client collection:(BUYCollection*)collection
{
    NSParameterAssert(client);
    self = [super initWithStyle:UITableViewStyleGrouped];
    if (self) {
        self.client = client;
        self.collection = collection;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];

    if (self.collection) {
    } else {
    }

    [self.tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:@"Cell"];
    UINib *nib = [UINib nibWithNibName:@"ProductListCell" bundle:nil];
    [self.tableView registerNib:nib forCellReuseIdentifier:@"ProductListCell"];

    if (self.collection) {
        // If we're presenting with a collection, add the ability to sort
        UIBarButtonItem *sortBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"Sort" style:UIBarButtonItemStylePlain target:self action:@selector(presentCollectionSortOptions:)];
        self.navigationItem.rightBarButtonItem = sortBarButtonItem;
    }

    [self loadProducts];
}

- (void) loadProducts {
    if (self.collection) {
        [self getCollectionWithSortOrder:BUYCollectionSortCollectionDefault];
    } else if (self.tags) {
        [self getCollectionByTags];
    } else {
        // todo implement loading all pages as user scrolls UITableView
        [[UIApplication sharedApplication] setNetworkActivityIndicatorVisible:YES];
        [self.client getProductsPage:1 completion:^(NSArray *products, NSUInteger page, BOOL reachedEnd, NSError *error) {
            [[UIApplication sharedApplication] setNetworkActivityIndicatorVisible:NO];
            if (error == nil && products) {
                self.products = [self filterProducts:products];
                [self.tableView reloadData];
            } else {
                NSLog(@"Error fetching products: %@", error);
            }
        }];
    }
}

- (NSArray *)filterProducts:(NSArray *)products {
    if (self.minPrice == nil && self.maxPrice == nil && self.searchKeyword == nil) {
        return products;
    }
    NSString *searchKeywordLowerCase = self.searchKeyword == nil || self.searchKeyword.length == 0 ? nil : [self.searchKeyword lowercaseString];
    NSMutableArray *result = [[NSMutableArray alloc] initWithCapacity:products.count];
    for (NSUInteger i = 0; i < products.count; i++) {
        BUYProduct *product = products[i];
        NSDecimalNumber *price = product.minimumPrice;
        if (price == nil) {
            continue;
        }
        if (self.minPrice != nil && [price compare:self.minPrice] < 0) {
            continue;
        }
        if (self.maxPrice != nil && [price compare:self.maxPrice] > 0) {
            continue;
        }
        if (searchKeywordLowerCase != nil && ![self product:product matchesKeyword:searchKeywordLowerCase]) {
            continue;
        }
        [result addObject:product];
    }
    return result;
}

- (BOOL)product:(BUYProduct *)product matchesKeyword:(NSString *)keyword {
    NSString *title = [product.title lowercaseString];
    NSString *description = [product.title lowercaseString];
    return [title rangeOfString:keyword].location != NSNotFound || [description rangeOfString:keyword].location != NSNotFound;
}

- (void)dealloc
{
    [self.collectionOperation cancel];
}

- (void)presentCollectionSortOptions:(id)sender
{
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Collection Sort" message:nil preferredStyle:UIAlertControllerStyleActionSheet];
    
    [alertController addAction:[UIAlertAction actionWithTitle:@"Default" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [self getCollectionWithSortOrder:BUYCollectionSortCollectionDefault];
    }]];
    [alertController addAction:[UIAlertAction actionWithTitle:@"Best Selling" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [self getCollectionWithSortOrder:BUYCollectionSortBestSelling];
    }]];
    [alertController addAction:[UIAlertAction actionWithTitle:@"Title - Ascending" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [self getCollectionWithSortOrder:BUYCollectionSortTitleAscending];
    }]];
    [alertController addAction:[UIAlertAction actionWithTitle:@"Title - Descending" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [self getCollectionWithSortOrder:BUYCollectionSortTitleDescending];
    }]];
    [alertController addAction:[UIAlertAction actionWithTitle:@"Price - Ascending" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [self getCollectionWithSortOrder:BUYCollectionSortPriceAscending];
    }]];
    [alertController addAction:[UIAlertAction actionWithTitle:@"Price - Descending" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [self getCollectionWithSortOrder:BUYCollectionSortPriceDescending];
    }]];
    
    /*  Note: The BUYCollectionSortCreatedAscending and BUYCollectionSortCreatedDescending are currently not support
    [alertController addAction:[UIAlertAction actionWithTitle:@"BUYCollectionSortCreatedAscending" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [self getCollectionWithSortOrder:BUYCollectionSortCreatedAscending];
    }]];
    [alertController addAction:[UIAlertAction actionWithTitle:@"BUYCollectionSortCreatedDescending" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [self getCollectionWithSortOrder:BUYCollectionSortCreatedDescending];
    }]];
     */
    
    if (self.traitCollection.userInterfaceIdiom == UIUserInterfaceIdiomPad) {
        UIPopoverPresentationController *popPresenter = [alertController popoverPresentationController];
        popPresenter.barButtonItem = (UIBarButtonItem*)sender;
    }
    
    [alertController addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:NULL]];
    
    [self presentViewController:alertController animated:YES completion:NULL];
}

- (void)getCollectionWithSortOrder:(BUYCollectionSort)collectionSort
{
    [self.collectionOperation cancel];
    [[UIApplication sharedApplication] setNetworkActivityIndicatorVisible:YES];
    self.collectionOperation = [self.client getProductsPage:1 inCollection:self.collection.identifier withTags:nil sortOrder:collectionSort completion:^(NSArray *products, NSUInteger page, BOOL reachedEnd, NSError *error) {
        [[UIApplication sharedApplication] setNetworkActivityIndicatorVisible:NO];
        
        if (error == nil && products) {
            self.products = [self filterProducts:products];
            [self.tableView reloadData];
        }
        else {
            NSLog(@"Error fetching products: %@", error);
        }
    }];
}

- (void)getCollectionByTags
{
    [self.collectionOperation cancel];
    [[UIApplication sharedApplication] setNetworkActivityIndicatorVisible:YES];

    id collectionId = self.collection == nil ? @0 : self.collection.identifier; // cannot pass nil to API, instead empty string should be passed

    self.collectionOperation = [self.client
            getProductsByTags:self.tags
                         page:1
                 completion:^(NSArray * _Nullable products, NSError * _Nullable error) {
                     [[UIApplication sharedApplication] setNetworkActivityIndicatorVisible:NO];
                     if (error == nil && products) {
                         self.products = [self filterProducts:products];
                         [self.tableView reloadData];
                     } else {
                         NSLog(@"Error fetching products: %@", error);
                     }
                 }];
}

#pragma mark - Table View

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.products.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    ProductListCell *cell = [tableView dequeueReusableCellWithIdentifier:@"ProductListCell" forIndexPath:indexPath];
    BUYProduct *product = self.products[indexPath.row];
    [cell setupWithProduct:product];
    return cell;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return ProductListCell.HEIGHT;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    BUYProduct *product = self.products[indexPath.row];
    [self.tableView deselectRowAtIndexPath:indexPath animated:YES];
    [self demoProductViewControllerWithProduct:product];
}

- (void)demoProductViewControllerWithProduct:(BUYProduct*)product
{
    ProductViewController *productViewController = [self productViewController];
    [productViewController loadWithProduct:product completion:^(BOOL success, NSError *error) {
        if (error == nil) {
            [self.navigationController pushViewController:productViewController animated:YES];

        }
    }];
}

-(ProductViewController*)productViewController
{
    ProductViewController *productViewController = [[ProductViewController alloc] initWithClient:self.client];
    productViewController.merchantId = MERCHANT_ID;
    return productViewController;
}

#pragma mark - UIViewControllerPreviewingDelegate
/*

-(UIViewController*)previewingContext:(id<UIViewControllerPreviewing>)previewingContext viewControllerForLocation:(CGPoint)location
{
    NSIndexPath *indexPath = [self.tableView indexPathForRowAtPoint:location];
    UITableViewCell *cell = [self.tableView cellForRowAtIndexPath:indexPath];
    if (cell == nil) {
        return nil;
    }
    
    BUYProduct *product = self.products[indexPath.row];
    ProductViewController *productViewController = [self productViewController];
    [productViewController loadWithProduct:product completion:NULL];
    
    previewingContext.sourceRect = cell.frame;
    
    return productViewController;
}

-(void)previewingContext:(id<UIViewControllerPreviewing>)previewingContext commitViewController:(UIViewController *)viewControllerToCommit
{
    [self presentViewController:viewControllerToCommit animated:YES completion:NULL];
}
*/

@end

File: Growler/ProductListViewController.swift
import Foundation
import UIKit
import Buy

class ProductListViewController: UITableViewController {

    var collection: BUYCollection?

    var collectionOperation: Operation?

    var products: [BUYProduct] = []

    var tags: [String]?

    var searchKeyword: String?

    var minPrice: NSDecimalNumber?

    var maxPrice: NSDecimalNumber?

    var pageIndex: Int = 0

    /// navController property is used if product list is inside of page controller.
    /// but we still needs to pop current controller when user taps close button
    var navController: UINavigationController?

    override func viewDidLoad() {
        super.viewDidLoad()
        self.tableView.backgroundColor = Colors.grayBackground
        self.tableView.contentInset = UIEdgeInsetsMake(8, 0, 0, 0) // according to mockups
        self.tableView.register(UITableViewCell.self, forCellReuseIdentifier: "Cell")
        let nib = UINib(nibName: "ProductListCell", bundle: nil)
        self.tableView.register(nib, forCellReuseIdentifier: "ProductListCell")
        self.loadProducts()
    }

    func loadProducts() {
        if self.collection != nil {
            self.getCollectionWithSortOrder(.collectionDefault)
        } else if self.tags != nil {
            self.getCollectionByTags()
        } else {
            // todo implement loading all pages as user scrolls UITableView
            ShopifyController.instance.client?.getProductsPage(1) {
                products, page, reachedEnd, error -> Void in
                if error == nil && products != nil {
                    self.products = self.filterProducts(products!)
                    self.tableView.reloadData()
                } else {
                    print("Error fetching products: \(error)")
                }
            }
        }
    }


    func filterProducts(_ products: [BUYProduct]) -> [BUYProduct] {
        if self.minPrice == nil && self.maxPrice == nil && self.searchKeyword == nil {
            return products
        }
        let searchKeywordLowerCase: String?
        if let searchKeyword = searchKeyword {
            if searchKeyword.isEmpty {
                searchKeywordLowerCase = nil
            } else {
                searchKeywordLowerCase = searchKeyword.lowercased()
            }
        } else {
            searchKeywordLowerCase = nil
        }
        var result = [BUYProduct]() /* capacity: products.count */
        for i in 0 ..< products.count {
            let product = products[i]
            guard let price = product.minimumPrice else {
                continue
            }
            if let minPrice = minPrice, price.compare(minPrice) == .orderedAscending {
                continue
            }
            if let maxPrice = maxPrice, price.compare(maxPrice) == .orderedDescending {
                continue
            }
            if let searchKeywordLowerCase = searchKeywordLowerCase, !doesProduct(product, matchKeyword: searchKeywordLowerCase) {
                continue
            }
            result.append(product)
        }
        return result
    }

    func doesProduct(_ product: BUYProduct, matchKeyword keyword: String) -> Bool {
        let title: String = product.title.lowercased()
        let description: String = product.title.lowercased()
        return (title as NSString).range(of: keyword).location != NSNotFound || (description as NSString).range(of: keyword).location != NSNotFound
    }

    deinit {
        self.collectionOperation?.cancel()
    }

    func getCollectionWithSortOrder(_ collectionSort: BUYCollectionSort) {
        collectionOperation?.cancel()
        guard let collection = collection else {
            return
        }
        self.collectionOperation = ShopifyController.instance.client.getProductsPage(1, inCollection: collection.identifier, withTags: nil, sortOrder: collectionSort) {
            products, page, reachedEnd, error -> Void in
            if error == nil {
                self.products = self.filterProducts(products ?? [])
                self.tableView.reloadData()
            } else {
                print("Error fetching products: \(error)")
            }
        }
    }

    func getCollectionByTags() {
        collectionOperation?.cancel()
        self.collectionOperation = ShopifyController.instance.client.getProductsByTags(tags ?? [], page: 1) {
            products, error -> Void in
            if error == nil {
                self.products = self.filterProducts(products ?? [])
                self.tableView.reloadData()
            } else {
                print("Error fetching products: \(error)")
            }
        }
    }

    override func numberOfSections(`in` tableView: UITableView) -> Int {
        return 2
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        switch section {
            case 0: return collection == nil ? 0 : 1
            default: return self.products.count
        }
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        switch indexPath.section {
            case 0:
                let cell = CollectionHeaderCell.loadFromNib()
                cell.productList = self
                cell.titleLabel.text = collection?.title
                cell.subtitleLabel.text = collection?.stringDescription
                return cell
            default:
                let cell = tableView.dequeueReusableCell(withIdentifier: "ProductListCell", for: indexPath) as! ProductListCell
                let product = products[indexPath.row]
                cell.setupWithProduct(product)
                return cell
        }
    }

    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        switch indexPath.section {
            case 0: return 77
            default: return ProductListCell.HEIGHT
        }
    }

    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        if indexPath.section == 1 {
            let product: BUYProduct? = self.products[indexPath.row]
            self.tableView.deselectRow(at: indexPath, animated: true)
            self.demoProductViewController(with: product!)
        }
    }

    func demoProductViewController(with product: BUYProduct) {
        let productViewController = ProductPageViewController(transitionStyle: .scroll, navigationOrientation: .horizontal, options: [:])
        productViewController.product = product
        self.navigationController?.pushViewController(productViewController, animated: true)
    }

}
